### PR TITLE
 Fix #5350: Fix purity of local lazy vals

### LINKED ISSUE #5350: Erasure removes Unit typed local lazy val references
```scala
scala> def foo = { lazy val bar: Unit = { println("Hello") }; bar }
def foo: Unit

scala> bar
// nothing happen
```

Given:
```scala
class Test {
  def foo = {
    lazy val bar: Unit = println("Hello")
    bar
  }
}
```
After erasure:
```scala
class Test() {
  def foo(): Unit = {
    lazy <accessor> def bar(): scala.runtime.BoxedUnit = {
      println("Hello")
      scala.runtime.BoxedUnit$#UNIT
    }
    () // no more call to `bar`
  }
}
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: tests/run/i5350.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+object Test {
+  def foo = {
+    lazy val bar: Unit = println("Hello")
```
Comment (nicolasstucki) [inline tests/run/i5350.scala]: @Blaisorblade does it make sense for this `lazy val` to be stable?

Comment (Blaisorblade) [reply tests/run/i5350.scala]: Yes. The definition of `isStable` doesn't check for `Lazy`, so `Symbol.isStable` should only imply idempotence.

To detect if a value is pure, Realizability checks `sym.isStable && !sym.is(Lazy | Erased, butNot = Module)`. Maybe here you should check
`sym.isStable && !sym.is(Lazy, butNot = Module)`? But on modules, this overlaps with #2266 and #4765.

That all needs testcases and consideration?

Comment (nicolasstucki) [reply tests/run/i5350.scala]: Actually modules do not reach [this line](https://github.com/lampepfl/dotty/pull/5703/files#diff-60732058b4c21c7127418de34cb149e9R403).  Hence the current condition does not affect the module init elimination.

File: compiler/src/dotty/tools/repl/Rendering.scala [hunk 2]
```diff
@@ -76,11 +76,8 @@ private[repl] class Rendering(parentClassLoader: Option[ClassLoader] = None) {
     val dcl = d.symbol.showUser
 
     try {
-      val resultValue =
-        if (d.symbol.is(Flags.Lazy)) Some("<lazy>")
-        else valueOf(d.symbol)
-
-      resultValue.map(value => s"$dcl = $value")
+      if (d.symbol.is(Flags.Lazy)) Some(dcl)
+      else valueOf(d.symbol).map(value => s"$dcl = $value")
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/repl/Rendering.scala]: :+1: on printing the flag to the left (in `ReplPrinter`). On this change, FWIW, Scala 2 also uses `<lazy>` here. Not sure anybody cares tho, so merging anyway.
```
scala> lazy val x = 1
x: Int = <lazy>
```

**Comment by Blaisorblade** [comment]
Tests confirm this doesn't affect the handling of modules, and that modules don't get to the changed code (because that code tests stability of functions, hence not of objects).

Hence merging.

**Pull request merged**